File: ai/rave.js
'use strict';

// BROWSER-COMPATIBLE: no Node.js-only APIs (require, process, etc.).
// Wrapped in an IIFE to avoid polluting the global namespace.
// Loaded as a plain <script> tag; do not add require/module/process at top level.

(function () {

/**
 * RAVE (Rapid Action Value Estimation) MCTS policy.
 *
 * Node structure: all stats kept in compact child-indexed arrays on the parent.
 * Child nodes are promoted lazily after N_EXPAND playout visits.
 */

const _isNode = typeof process !== 'undefined' && process.versions && process.versions.node;

const performance = (typeof window !== 'undefined') ? window.performance
  : require('perf_hooks').performance;

const { PASS, BLACK, WHITE } = _isNode ? require('../game2.js') : window.Game2;
const Util = _isNode ? require('../util.js') : window.Util;

const EXPLORATION_C = Util.envFloat('EXPLORATION_C', 1.4);

// Equivalence parameter.  Override with RAVE_EQUIV=<n>.
const RAVE_EQUIV = Util.envFloat('RAVE_EQUIV', 300);

// Fixed playout count per decision.  When non-zero, overrides the time budget.
const PLAYOUTS = Util.envInt('PLAYOUTS', 0);

// Minimum playout visits before a child node is promoted (allocated).
const N_EXPAND = Util.envInt('N_EXPAND', 3);

// ── Fast playout helpers ──────────────────────────────────────────────────────

// Random playout using Game2.  Records the flat cell indices of every non-pass
// move made by each player.  Returns { winner, blackPlayed, whitePlayed }.
function playTracked(game2) {
  const wasAlreadyOver = game2.gameOver;
  const N     = game2.N;
  const cap   = N * N;
  const cells = game2.cells;
  const nbr   = game2._nbr;
  const blackPlayed = [];
  const whitePlayed = [];

  const empty = [];
  for (let i = 0; i < cap; i++) {
    if (cells[i] === 0) empty.push(i);
  }

  const moveLimit = empty.length + 20;
  let moves = 0;

  while (!game2.gameOver && moves < moveLimit) {
    let placed = false;
    let end = empty.length;
    const current = game2.current;

    while (end > 0) {
      const ri  = Math.floor(Math.random() * end);
      const idx = empty[ri];
      empty[ri] = empty[end - 1];
      empty[end - 1] = idx;
      end--;

      if (game2.isTrueEye(idx)) continue;
      if (!game2.isLegal(idx))  continue;

      if (current === BLACK) blackPlayed.push(idx);
      else                    whitePlayed.push(idx);

      // Snapshot neighbour occupancy to detect captures after play.
      const base = idx * 4;
      const n0 = cells[nbr[base]], n1 = cells[nbr[base + 1]],
            n2 = cells[nbr[base + 2]], n3 = cells[nbr[base + 3]];
      game2.play(idx);
      empty[end] = empty[empty.length - 1];
      empty.pop();

      // If any previously occupied neighbour became empty, captures occurred.
      if ((n0 && !cells[nbr[base]])     || (n1 && !cells[nbr[base + 1]]) ||
          (n2 && !cells[nbr[base + 2]]) || (n3 && !cells[nbr[base + 3]])) {
        empty.length = 0;
        for (let i = 0; i < cap; i++) if (cells[i] === 0) empty.push(i);
      }

      placed = true;
      moves++;
      break;
    }

    if (!placed) { game2.play(PASS); moves++; }
  }

  let winner;
  if (wasAlreadyOver) {
    winner = game2.calcWinner();
  } else {
    winner = game2.estimateWinner();
  }
  return { winner, blackPlayed, whitePlayed };
}

// ── Tree node ─────────────────────────────────────────────────────────────────

// Enumerate legal non-true-eye moves from a Game2 state as integers.
// Place moves are y*N+x; pass is PASS (-1).
function getLegalMoves(game2) {
  const N     = game2.N;
  const cap   = N * N;
  const cells = game2.cells;
  const moves = [];
  for (let i = 0; i < cap; i++) {
    if (cells[i] !== 0) continue;
    if (game2.isTrueEye(i)) continue;
    if (game2.isLegal(i)) moves.push(i);
  }
  // Pass move must be at the end (if present).
  if (moves.length < cap / 3 || game2.consecutivePasses > 0) {
    moves.push(PASS);
  }
  return moves;
}

// Create a node for the position reached by `move`.
// `game2` is the game state AFTER `move` was played (or initial state for root).
// `ci`    is this node's index in parent.children / parent.child* arrays (-1 for root).
// `mover` is the player who played `move` (null at root).
// Priors (ladder + pattern) are computed eagerly from the current game state.
function makeNode(move, parent, ci, mover, game2, N) {
  const movesArr = getLegalMoves(game2);
  const M = movesArr.length;

  const legalMoves = new Int32Array(M);
  for (let i = 0; i < M; i++) legalMoves[i] = movesArr[i];

  const children    = new Array(M).fill(null);
  const wins        = new Float32Array(M).fill(0.01);
  const visits      = new Float32Array(M).fill(0.02);
  const raveWins    = new Float32Array(N * N).fill(0.01);
  const raveVisits  = new Float32Array(N * N).fill(0.02);

  return {
    move,
    parent,
    ci,           // this node's index in parent.children / parent.child* arrays (-1 for root)
    mover,        // player who made `move` to reach this node (null at root)
    totalVisits:  1,  // sum of visits; incremented each playout
    selectedChild: -1,  // set by selectAndExpand; read by backpropagate

    legalMoves,   // Int32Array(M)
    children,     // Array(M) — promoted child node or null

    wins,    // Float32Array(M) — playout wins per child
    visits,  // Float32Array(M) — playout visits per child

    raveWins,     // Float32Array(N*N) — RAVE wins indexed by cell; updated by rollouts
    raveVisits
  };
}

// RAVE-blended UCT score for child index i of node.                                                                                                                                                         
// Children with no real playout visits (cv === 0) get a large bonus so they                                                                                                                                 
// are always preferred over visited children.  RAVE (seeded with pattern                                                                                                                                    
// priors) ranks them within the unvisited tier.                                                                                                                                                             
// A separate priorBonus term decays as bonus/(1+realV), so ladder priors                                                                                                                                    
// guide early exploration without ever diluting the RAVE statistics.                                                                
function raveScore(moveIdx, node) {
  const move   = node.legalMoves[moveIdx];
  const realW  = node.wins[moveIdx];
  const realV  = node.visits[moveIdx];
  const raveW  = move === PASS ? 0 : node.raveWins[move];
  const raveV  = move === PASS ? 1 : node.raveVisits[move];
  const raveWR = raveW / raveV;
  const wr = (realW + raveWR * RAVE_EQUIV) / (realV + RAVE_EQUIV);
  if (realV < 1) {
    return 10 + wr + 0.001 * Math.random();
  }
  return 0.001 * Math.random() +
    wr +
    EXPLORATION_C * Math.sqrt(Math.log(node.totalVisits) / realV);
}

// ── RAVE-MCTS core ────────────────────────────────────────────────────────────

function selectAndExpand(root, rootGame2, N) {
  let node = root;
  const game2 = rootGame2.clone();

  while (!game2.gameOver) {
    const M = node.legalMoves.length;
    if (M === 0) break;

    // Select best child by RAVE-blended score.
    let best = 0, bestScore = -Infinity;
    for (let i = 0; i < M; i++) {
      const s = raveScore(i, node);
      if (s > bestScore) { bestScore = s; best = i; }
    }

    const mover = game2.current;   // player making this move
    game2.play(node.legalMoves[best]);

    // Promote child to a full node once it has accumulated enough visits.
    // Fall through to the descent check so the loop continues into the new node;
    // its children all have cv=0 < N_EXPAND, so the leaf case fires next iteration
    // (exactly one makeNode per playout, same as rave2's one-expansion-per-playout).
    if (node.children[best] === null && node.visits[best] >= N_EXPAND) {
      node.children[best] = makeNode(node.legalMoves[best], node, best, mover, game2, N);
    }

    // After a pass, always force a second pass so the playout scores the current
    // board position (consecutive passes end the game).  This prevents rollouts
    // from a single-pass state playing on for many random moves and inflating
    // the pass move's apparent win rate.
    if (!game2.gameOver && game2.consecutivePasses > 0) {
      game2.play(PASS);
      node.selectedChild = best;
      break;
    }

    // Descend into the promoted child, if available.
    if (node.children[best] !== null) {
      node = node.children[best];
      node.selectedChild = -1;  // reset in case game ends before we select below
      continue;
    }

    // Unpromoted leaf — run playout from here.
    node.selectedChild = best;
    break;
  }

  return { node, game2 };
}

// Backpropagate playout result and update RAVE statistics.
//
// node.selectedChild holds the unpromoted child index that was played last,
// or -1 if we descended all the way to a promoted node (game already over).
//
// childMover(n): the player choosing the next move from node n.
//   - rootPlayer if n is the root (n.mover === null)
//   - opponent(n.mover) otherwise
function backpropagate(node, winner, blackPlayed, whitePlayed, rootPlayer) {
  function childMover(n) {
    return n.mover === null ? rootPlayer : (n.mover === BLACK ? WHITE : BLACK);
  }

  function updateRave(node, won, played) {
    const invPlayedCount = 1 / (1 + played.length)
    let weight = 1;
    for (let k = 0; k < played.length; k++) {
      const move = played[k];
      node.raveVisits[move] += weight;
      node.raveWins[move] += won * weight;
      weight -= invPlayedCount;
    }
  }

  // Update the unpromoted leaf child stats (if we stopped before descending).
  // Also update RAVE at this node so root's RAVE is populated even when no
  // deeper promoted nodes exist (e.g. N_EXPAND=9999).
  const leafIdx = node.selectedChild;
  if (leafIdx !== -1) {
    const chooser = childMover(node);
    const won     = winner === chooser ? 1 : 0;
    node.visits[leafIdx]++;
    node.wins[leafIdx] += won;
    node.totalVisits++;

    const played = chooser === BLACK ? blackPlayed : whitePlayed;
    updateRave(node, won, played);
  }

  // Walk up the tree, updating each parent's child arrays and RAVE arrays.
  while (node.parent !== null) {
    const ci  = node.ci;   // stored at promotion time — no lookup needed
    const won = winner === node.mover ? 1 : 0;
    node.parent.visits[ci]++;
    node.parent.wins[ci] += won;
    node.parent.totalVisits++;

    // RAVE update at node.parent: credit the parent's chooser's playout moves.
    const chooser = childMover(node.parent);
    const played  = chooser === BLACK ? blackPlayed : whitePlayed;
    const rWon    = winner === chooser ? 1 : 0;
    updateRave(node.parent, rWon, played);

    node = node.parent;
  }
}

// ── Public interface ──────────────────────────────────────────────────────────

function getMove(game, timeBudgetMs) {
  if (game.gameOver) return { type: 'pass', info: 'game already over' };

  const N          = game.cells ? game.N : game.boardSize;
  const game2      = game.cells ? game.clone() : game.toGame2();
  const rootPlayer = game2.current;

  // Obvious pass: opponent just passed and we're already winning — end the game.
  if (game2.consecutivePasses > 0 && game2.calcWinner() === rootPlayer) {
    return { type: 'pass', info: 'obvious pass: already winning', rootWinRatio: 1 };
  }

  const root = makeNode(null, null, -1, null, game2, N);

  const deadline = performance.now() + timeBudgetMs;
  let playouts = 0;

  do {
    playouts++;
    const { node, game2: simGame2 } = selectAndExpand(root, game2, N);
    const { winner, blackPlayed, whitePlayed } = playTracked(simGame2);
    backpropagate(node, winner, blackPlayed, whitePlayed, rootPlayer);
  } while (PLAYOUTS > 0 ? playouts < PLAYOUTS : performance.now() < deadline);

  // Best child: most playout visits; ties broken by RAVE-blended score.
  const M = root.legalMoves.length;
  let bestIdx = 0, bestVisits = -1, bestScore = -Infinity;
  for (let i = 0; i < M; i++) {
    const cv = root.visits[i];
    if (cv > bestVisits || (cv === bestVisits && raveScore(i, root) > bestScore)) {
      bestVisits = cv;
      bestScore  = raveScore(i, root);
      bestIdx    = i;
    }
  }

  const children = [];
  for (let i = 0; i < M; i++) {
    const m = root.legalMoves[i];
    children.push({
      move: m === PASS ? { type: 'pass' } : { type: 'place', x: m % N, y: (m / N) | 0 },
      visits: root.visits[i],
      wins:   root.wins[i],
    });
  }
  children.sort((a, b) => b.visits - a.visits);

  let totalChildWins = 0;
  for (let i = 0; i < M; i++) totalChildWins += root.wins[i];
  const rootWinRatio = root.totalVisits > 0 ? totalChildWins / root.totalVisits : 0.5;

  if (root.wins[bestIdx] === 0 && game.moveCount >= N * N / 2) {
    return { type: 'pass', info: 'no winning line found', children, rootWinRatio };
  }

  const m = root.legalMoves[bestIdx];
  const cv = root.visits[bestIdx];
  const bestWinRatio = cv > 0 ? root.wins[bestIdx] / cv : 0.5;

  const result = m === PASS ? { type: 'pass', children, rootWinRatio }
                            : { type: 'place', x: m % N, y: (m / N) | 0, children, rootWinRatio };
  result.info = `win likelihood: ${bestWinRatio.toFixed(3)}`;
  return result;
}

if (typeof module !== 'undefined') module.exports = getMove;
else window.getMove = getMove;

})();

File: c/rave.c
/*
 * rave.c — RAVE-MCTS engine (C port of ai/rave.js).
 */
#include "rave.h"
#include <stdlib.h>
#include <string.h>
#include <float.h>
#include <time.h>
#include <stdio.h>

/* ── Node allocation (chunked) ─────────────────────────────────────────────── */

static RaveNode *alloc_node(RaveState *s) {
    if (s->cur_used >= RAVE_CHUNK_SIZE) {
        /* Need a new chunk — reuse next in chain or allocate */
        if (s->cur->next == NULL) {
            RaveChunk *c = malloc(sizeof(RaveChunk));
            if (!c) { fprintf(stderr, "rave: out of memory\n"); exit(1); }
            c->next = NULL;
            s->cur->next = c;
        }
        s->cur = s->cur->next;
        s->cur_used = 0;
    }
    RaveNode *n = &s->cur->nodes[s->cur_used++];
    s->total_used++;
    memset(n, 0, sizeof(RaveNode));
    return n;
}

/* ── Get legal moves ───────────────────────────────────────────────────────── */

static int get_legal_moves(const Game2 *g, int32_t *out) {
    int n = 0;
    for (int i = 0; i < CAP; i++) {
        if (g->cells[i] != EMPTY) continue;
        if (g2_is_true_eye(g, i)) continue;
        if (g2_is_legal(g, i)) out[n++] = i;
    }
    if (n < CAP / 3 || g->consecutive_passes > 0)
        out[n++] = PASS;
    return n;
}

/* ── Make node ─────────────────────────────────────────────────────────────── */

static RaveNode *make_node(RaveState *s, int32_t move, RaveNode *parent,
                           int32_t ci, const Game2 *game) {
    RaveNode *n = alloc_node(s);

    n->move   = move;
    n->parent = parent;
    n->ci     = ci;
    n->mover  = -game->current;  /* player who made `move` */
    n->total_visits = 0.1f;
    n->selected_child = -1;

    n->num_moves = get_legal_moves(game, n->legal_moves);
    int M = n->num_moves;

    for (int i = 0; i < M; i++) {
        n->children[i] = NULL;
        n->wins[i]     = RAVE_PRIOR_WINS;
        n->visits[i]   = RAVE_PRIOR_VISITS;
    }

    if (parent == NULL || parent->parent == NULL) {
        for (int i = 0; i < CAP; i++) {
            n->rave_wins[i]   = RAVE_PRIOR_WINS;
            n->rave_visits[i] = RAVE_PRIOR_VISITS;
        }
    } else {
        const RaveNode *gp = parent->parent;
        for (int i = 0; i < CAP; i++) {
            n->rave_wins[i]   = RAVE_INHERIT * gp->rave_wins[i];
            n->rave_visits[i] = RAVE_INHERIT * gp->rave_visits[i];
        }
    }

    return n;
}

/* ── UCB score ─────────────────────────────────────────────────────────────── */

static float ucb_score(int move_idx, const RaveNode *node) {
    int32_t move = node->legal_moves[move_idx];

    float rave_wr = (move == PASS) ? 0.0f
        : (node->rave_wins[move] / node->rave_visits[move]);

    float real_w  = node->wins[move_idx];
    float real_v  = node->visits[move_idx];
    float real_wr = real_w / real_v;

    float rave_weight = RAVE_K / (RAVE_K + real_v);
    float wr = (1.0f - rave_weight) * real_wr + rave_weight * rave_wr;

    return wr + 0.001f * g2_randf()
         + RAVE_EXPLORATION_C * sqrtf(logf(node->total_visits) / real_v);
}

/* ── Playout (tracked) ─────────────────────────────────────────────────────── */

static int8_t play_tracked(Game2 *g, float *played) {
    memset(played, 0, sizeof(float) * CAP);

    int move_limit = 3 * g->empty_count + 20;
    float weight_step = 1.0f / CAP;
    int moves = 0;
    float weight = 1.0f;

    while (!g->game_over && moves < move_limit) {
        int8_t current = g->current;
        int32_t idx = g2_random_legal_move(g);
        if (idx == PASS) { g2_play(g, PASS); moves++; continue; }
        if (played[idx] == 0.0f)
            played[idx] = (current == BLACK) ? weight : -weight;
        g2_play(g, idx);
        moves++;
        weight -= weight_step;
    }

    return g2_estimate_winner(g);
}

/* ── Select and expand ─────────────────────────────────────────────────────── */

typedef struct {
    RaveNode *node;
    Game2     game;
} SelectResult;

static SelectResult select_and_expand(RaveState *s, RaveNode *root, const Game2 *root_game) {
    RaveNode *node = root;
    Game2 game;
    g2_clone(&game, root_game);

    while (!game.game_over) {
        int M = node->num_moves;
        if (M == 0) break;

        /* Select best child */
        int best = 0;
        float best_score = -FLT_MAX;
        for (int i = 0; i < M; i++) {
            float sc = ucb_score(i, node);
            if (sc > best_score) { best_score = sc; best = i; }
        }

        g2_play(&game, node->legal_moves[best]);

        /* Promote if enough visits */
        if (node->children[best] == NULL && node->visits[best] >= RAVE_N_EXPAND)
            node->children[best] = make_node(s, node->legal_moves[best], node, best, &game);

        /* Force second pass to end game */
        if (!game.game_over && game.consecutive_passes > 0) {
            g2_play(&game, PASS);
            node->selected_child = best;
            break;
        }

        /* Descend into promoted child */
        if (node->children[best] != NULL) {
            node = node->children[best];
            node->selected_child = -1;
            continue;
        }

        /* Unpromoted leaf */
        node->selected_child = best;
        break;
    }

    return (SelectResult){ node, game };
}

/* ── Backpropagate ─────────────────────────────────────────────────────────── */

static void update_rave(RaveNode *node, float won, const float *played, int8_t chooser) {
    if (chooser == BLACK) {
        for (int k = 0; k < CAP; k++) {
            float w = played[k];
            if (w > 0.0f) {
                node->rave_visits[k] += w;
                node->rave_wins[k]   += won * w;
            }
        }
    } else {
        for (int k = 0; k < CAP; k++) {
            float w = played[k];
            if (w < 0.0f) {
                node->rave_visits[k] -= w;
                node->rave_wins[k]   -= won * w;
            }
        }
    }
}

static void backpropagate(RaveNode *node, int8_t winner, const float *played) {
    /* Update leaf child stats */
    int leaf = node->selected_child;
    if (leaf != -1) {
        int8_t chooser = -node->mover;
        float won = (winner == chooser) ? 1.0f : 0.0f;
        node->visits[leaf]++;
        node->wins[leaf] += won;
        node->total_visits++;
        update_rave(node, won, played, chooser);
    }

    /* Walk up the tree */
    while (node->parent != NULL) {
        int ci = node->ci;
        int8_t chooser = -node->parent->mover;
        float won = (winner == chooser) ? 1.0f : 0.0f;
        node->parent->visits[ci]++;
        node->parent->wins[ci] += won;
        node->parent->total_visits++;
        update_rave(node->parent, won, played, chooser);
        node = node->parent;
    }
}

/* ── Public API ────────────────────────────────────────────────────────────── */

RaveState *rave_create(void) {
    RaveState *s = malloc(sizeof(RaveState));
    if (!s) { fprintf(stderr, "rave: out of memory\n"); exit(1); }
    RaveChunk *c = malloc(sizeof(RaveChunk));
    if (!c) { fprintf(stderr, "rave: out of memory\n"); exit(1); }
    c->next = NULL;
    s->head = c;
    s->cur = c;
    s->cur_used = 0;
    s->total_used = 0;
    return s;
}

void rave_destroy(RaveState *s) {
    if (!s) return;
    RaveChunk *c = s->head;
    while (c) { RaveChunk *next = c->next; free(c); c = next; }
    free(s);
}

RaveResult rave_search(RaveState *s, const Game2 *root_game,
                       int playout_limit, int time_ms) {
    /* Reset pool — reuse existing chunks, no free/alloc */
    s->cur = s->head;
    s->cur_used = 0;
    s->total_used = 0;

    if (root_game->game_over)
        return (RaveResult){ PASS, 0.5f, 0 };

    /* Obvious pass: opponent just passed and we're winning */
    if (root_game->consecutive_passes > 0 &&
        g2_estimate_winner(root_game) == root_game->current)
        return (RaveResult){ PASS, 1.0f, 0 };

    RaveNode *root = make_node(s, -1, NULL, -1, root_game);

    int playouts = 0;
    if (playout_limit > 0) {
        for (int p = 0; p < playout_limit; p++) {
            SelectResult sr = select_and_expand(s, root, root_game);
            int8_t winner = play_tracked(&sr.game, s->played);
            backpropagate(sr.node, winner, s->played);
            playouts++;
        }
    } else {
        clock_t deadline = clock() + (clock_t)((double)time_ms / 1000.0 * CLOCKS_PER_SEC);
        do {
            SelectResult sr = select_and_expand(s, root, root_game);
            int8_t winner = play_tracked(&sr.game, s->played);
            backpropagate(sr.node, winner, s->played);
            playouts++;
        } while (clock() < deadline);
    }

    /* Best child: most visits, ties broken by UCB score */
    int M = root->num_moves;
    int best_idx = 0;
    float best_visits = -1.0f, best_score = -FLT_MAX;
    for (int i = 0; i < M; i++) {
        float cv = root->visits[i];
        if (cv > best_visits || (cv == best_visits && ucb_score(i, root) > best_score)) {
            best_visits = cv;
            best_score  = ucb_score(i, root);
            best_idx    = i;
        }
    }

    float total_wins = 0;
    for (int i = 0; i < M; i++) total_wins += root->wins[i];
    float win_ratio = total_wins / root->total_visits;

    /* Give up if no winning line in late game. Only reliable with enough playouts. */
    if (playouts >= RAVE_RESIGN_MIN_PLAYOUTS &&
        root_game->empty_count <= CAP / 2 &&
        root->wins[best_idx] <= RAVE_PRIOR_WINS)
        return (RaveResult){ PASS, win_ratio, playouts };

    return (RaveResult){ root->legal_moves[best_idx], win_ratio, playouts };
}
